### ISSUE TITLE
Yolo Occasionally Misses Text Lines in Document Detection Task

### ISSUE THREAD

**Issue by alymsalahin3** [ultralytics/ultralytics#17839]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

## Description
I am training a two-class model on a custom dataset to detect text and tables within documents. However, I am encountering a recurring issue where the model occasionally fails to detect certain text lines. Below is a summary of the issue:

## Observations
**Issue**: Some text lines are not detected at all, even though they are present in the training dataset.

**Patterns**:

- The issue occurs inconsistently, arising even on images that are highly similar to those where the model performs well.

- The only consistent observation is that the missed text typically consists of single lines rather than entire paragraphs. Missing whole paragraphs is very rare.

- Sometimes, similar single lines are detected; however, the bounding boxes are smaller than expected and accompanied by low confidence scores, as illustrated in the samples below.

### 1) Missed Lines
![Pasted image](https://github.com/user-attachments/assets/83e235e1-b048-415a-8e83-6f872372a3ee)
![Pasted image (4)](https://github.com/user-attachments/assets/4d514c32-7e01-4249-8506-e1b4c6951a30)
![Pasted image (3)](https://github.com/user-attachments/assets/bfd74c86-3ab9-46cb-829a-f1dcc88b4858)

### 2) Detected but smaller than expected:
![2024-07-22-00-17-13-335744-169282830444444940724509271206412953305_LIMID Phase II Guidelines_32163](https://github.com/user-attachments/assets/610e932d-a248-4dc4-b47d-4b7ce890c0e1)

### 3) Rarely missing paragraphs (multiple lines):

![Pasted image (5)](https://github.com/user-attachments/assets/8520f99e-3836-4e7c-9189-eac657407ae9)

### Data Validation

- Used data exploration tools to ensure annotation consistency and variety in the dataset.
- Clustered data to identify any potential gaps.

### Augmentation Verification
- Verified that augmentations do not cause any bounding box loss during transformations.

### Model Adjustments
- Switched from YOLOv8x to YOLOv10x to address potential NMS issues.
- Experimented with fine-tuning parameters and augmentation strategies.

### Data Scaling
- Gradually increased the training dataset size from 4k to 12k images, focusing on samples with low average confidence. While this improved general performance, the issue of missed text lines persists.

**Are there any probable causes for this issue or additional recommendations to resolve it?**

**Comment by UltralyticsAssistant** [comment]
👋 Hello @alymsalahin3, thank you for your detailed report and for using Ultralytics 🚀! We appreciate the time you’ve taken to provide observations and samples related to your custom detection task.

We recommend visiting the [Ultralytics Docs](https://docs.ultralytics.com) for additional training insights, including [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) that might be helpful for fine-tuning your model. If you're encountering a 🐛 Bug, please provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us investigate further.

Here are some initial troubleshooting steps to ensure optimal performance:  
## Upgrade  
Ensure you are using the latest `ultralytics` release and dependencies, as improvements are continuously added to address known issues. Update with:  
```bash  
pip install -U ultralytics  
```  

## Environments  
YOLO can be run in various verified environments with preinstalled dependencies:  
- **Notebooks** with free GPU:  
  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a>  
  <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>  
  <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>  
- **Cloud Platforms** like Google Cloud (see [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)) or AWS (see [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)).  
- [Docker](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/): <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>  

## Recommendations  
Below are suggestions for your use case:  
1. **Analyze Missed Detections**: Review the missed or low-confidence cases in your dataset and consider whether annotations include variability, such as single-line text in different positions or resolutions.  
2. **Augmentation Improvements**: Revisit augmentation parameters to ensure they provide adequate variability without distorting single-line text.  
3. **Adjust Anchor Sizes**: Since single-line text is smaller in scale, manually optimizing anchor sizes for your dataset might improve NMS performance.  
4. **Dataset Analysis**: Validate your data format and annotations for consistency (e.g., no missing or duplicate bounding boxes). YOLO models are annotation-sensitive, and inconsistencies could impact performance, especially for edge cases.  

Please share additional information (e.g., training logs, hyperparameters, or a small dataset snippet for reproducibility) if the problem persists. This will help us assist you more effectively! 😊  

For a vibrant discussion or live help, feel free to join the [Ultralytics Discord](https://discord.com/invite/ultralytics) 🎧, share your experience on our [Subreddit](https://reddit.com/r/Ultralytics), or post on [Discourse](https://community.ultralytics.com).

## Status  
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>  
This indicates whether our CI system has passed the latest integration tests. If green, YOLO modes are working as expected on macOS, Windows, and Ubuntu.  

This is an automated response to help address your issue promptly 😊. An Ultralytics engineer will review and assist further soon!

**Comment by Y-T-G** [comment]
You can try P6 model.

https://github.com/ultralytics/ultralytics/issues/16895#issuecomment-2413016050

**Comment by Y-T-G** [comment]
What imgsz are you using?

**Comment by alymsalahin3** [comment]
Could this potentially result in the loss of important features, especially considering that my dataset contains large images?

Additionally, would you recommend adjusting the anchor boxes, as suggested here (https://github.com/ultralytics/ultralytics/issues/10229#issuecomment-2075427078), given the diversity of objects in my dataset?

**Comment by Y-T-G** [comment]
You can run prediction at 320 or 480 without retraining.

**Comment by alymsalahin3** [comment]
Hey @Y-T-G , Trying a smaller `imgsz` didn’t yield better results; instead, it missed more objects. One thing I tried that showed significant improvement for most (though not all) samples was using YOLOv11 with `reg_max=1.` However, while the results improved a lot, the issue still persists for a **fewer samples.**

Do you suggest trying anything else that might help enhance the results?
Also, would adding a P6 head to the YOLOv11 config might help enhancing the model's recall and ability to detect better?

**Comment by Y-T-G** [comment]
You can try. P6 heads should have higher receptive field.

**Comment by Y-T-G** [comment]
Can you provide the output after running this command in terminal: `yolo checks`?

**Comment by alymsalahin3** [comment]
sure, below you can find the command's output:
```shell
/opt/conda/lib/python3.11/site-packages/albumentations/__init__.py:24: UserWarning: A new version of Albumentations is available: 1.4.23 (you have 1.4.22). Upgrade using: pip install -U albumentations. To disable automatic update checks, set the environment variable NO_ALBUMENTATIONS_UPDATE to 1.
  check_for_updates()
Ultralytics 8.3.49 🚀 Python-3.11.10 torch-2.5.1+cu124 CUDA:0 (NVIDIA GeForce RTX 3090, 24253MiB)
Setup complete ✅ (128 CPUs, 503.5 GB RAM, 15.9/61.0 GB disk)

OS                  Linux-5.15.0-113-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.11.10
Install             pip
RAM                 503.52 GB
Disk                15.9/61.0 GB
CPU                 AMD EPYC 7B13 64-Core Processor
CPU count           128
GPU                 NVIDIA GeForce RTX 3090, 24253MiB
GPU count           1
CUDA                12.4

numpy               ✅ 2.1.2>=1.23.0
numpy               ✅ 2.1.2<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.10.0>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 11.0.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.5.1+cu124>=1.8.0
torch               ✅ 2.5.1+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.20.1+cu124>=0.9.0
tqdm                ✅ 4.66.5>=4.64.0
psutil              ✅ 6.1.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.13>=2.0.0
```

**Comment by Y-T-G** [comment]
Maybe you can use `close_mosaic=0`

**Comment by alymsalahin3** [comment]
please note I am currently training without mosaic augmentation, since I noticed it sometimes drop the label box when a big object (big paragraph or table) is cropped, so I end up with unlabeled text.
```
yolo train model=yolo11x.yaml data=data.yaml epochs=300 imgsz=640 batch=16  project=layout/runs degrees=2.0 perspective=0.0001 translate=0.0 erasing=0.0 mosaic=0.0 optimizer=AdamW lr0=0.001 lrf=0.00001
```

**Comment by Y-T-G** [comment]
`lrf` is not final learning rate. It's learning rate factor. `final_lr = lr0 * lrf`.

**Comment by alymsalahin3** [comment]
Although I adjusted it, the problem still persists. I even tried simplifying the P6 head as follows:
```
- [-1, 1, Conv, [1024, 3, 2]]
- [[-1, 12], 1, Concat, [1]] # cat head P6
- [-1, 1, C3k2, [1024, True]
``` 
However, the metrics still drop rapidly after some epochs. Should I remove the P6? If yes, do you suggest trying anything else that might enhance the model's detection and fix the missing objects issue?

**Comment by Y-T-G** [comment]
You can try v8-p6 model

**Comment by alymsalahin3** [comment]
I've got the same issue with the v8-p6 model too

**Comment by glenn-jocher** [comment]
Consider reviewing your dataset and annotations for potential inconsistencies or edge cases. If the issue persists, you might want to experiment with transfer learning from a model pretrained on a similar task or adjust your training hyperparameters further.

**Comment by vogelbam** [comment]
@alymsalahin3 have you found a solution to this problem? I am working with totally different inputs but I'm observing similar issues.
